### ISSUE TITLE
[Bug]: Not able to make call to Ollama multimodal model in the cookbook

### ISSUE THREAD

**Issue by AutoAPEAI** [run-llama/llama_index#10490]
### Bug Description

When I trying to follow the steps int the [cookbook](https://docs.llamaindex.ai/en/latest/examples/multi_modal/ollama_cookbook.html#), I was getting an error saying that there's a ResponseError: calling Ollama. However, I have installed Ollama successfully and was able to make model call via curl command. I suspect that this has something to do with http://localhost:11434.

### Version

llama-index 0.9.44

### Steps to Reproduce

Following the steps in the [cookbook](https://docs.llamaindex.ai/en/latest/examples/multi_modal/ollama_cookbook.html#)
```
from pydantic import BaseModel

class Restaurant(BaseModel):
    """Data model for an restaurant."""

    restaurant: str
    food: str
    discount: str
    price: str
    rating: str
    review: str

from llama_index.program import MultiModalLLMCompletionProgram
from llama_index.output_parsers import PydanticOutputParser

prompt_template_str = """\
{query_str}

Return the answer as a Pydantic object. The Pydantic schema is given below:

"""
mm_program = MultiModalLLMCompletionProgram.from_defaults(
    output_parser=PydanticOutputParser(Restaurant),
    image_documents=image_documents,
    prompt_template_str=prompt_template_str,
    multi_modal_llm=mm_model,
    verbose=True,
)

response = mm_program(query_str="Can you summarize what is in the image?")
for res in response:
    print(res)
```

### Relevant Logs/Tracbacks

{
	"name": "ResponseError",
	"message": "",
	"stack": "---------------------------------------------------------------------------
ResponseError                             Traceback (most recent call last)
Cell In[13], line 1
----> 1 response = mm_program(query_str=\"Can you summarize what is in the image?\")
      2 for res in response:
      3     print(res)

File /usr/local/lib/python3.11/site-packages/llama_index/program/multi_modal_llm_program.py:85, in MultiModalLLMCompletionProgram.__call__(self, llm_kwargs, *args, **kwargs)
     82 llm_kwargs = llm_kwargs or {}
     83 formatted_prompt = self._prompt.format(llm=self._multi_modal_llm, **kwargs)
---> 85 response = self._multi_modal_llm.complete(
     86     formatted_prompt,
     87     image_documents=self._image_documents,
     88     **llm_kwargs,
     89 )
     91 raw_output = response.text
     92 if self._verbose:

File /usr/local/lib/python3.11/site-packages/llama_index/multi_modal_llms/ollama.py:158, in OllamaMultiModal.complete(self, prompt, image_documents, formatted, **kwargs)
    155 \"\"\"Complete.\"\"\"
    156 import ollama
--> 158 response = ollama.generate(
    159     model=self.model,
    160     prompt=prompt,
    161     images=image_documents_to_base64(image_documents),
    162     stream=False,
    163     options=self._model_kwargs,
    164 )
    165 return CompletionResponse(
    166     text=response[\"response\"],
    167     raw=response,
    168     additional_kwargs=get_additional_kwargs(response, (\"response\",)),
    169 )

File /usr/local/lib/python3.11/site-packages/ollama/_client.py:111, in Client.generate(self, model, prompt, system, template, context, stream, raw, format, images, options, keep_alive)
    108 if not model:
    109   raise RequestError('must provide a model')
--> 111 return self._request_stream(
    112   'POST',
    113   '/api/generate',
    114   json={
    115     'model': model,
    116     'prompt': prompt,
    117     'system': system,
    118     'template': template,
    119     'context': context or [],
    120     'stream': stream,
    121     'raw': raw,
    122     'images': [_encode_image(image) for image in images or []],
    123     'format': format,
    124     'options': options or {},
    125     'keep_alive': keep_alive,
    126   },
    127   stream=stream,
    128 )

File /usr/local/lib/python3.11/site-packages/ollama/_client.py:82, in Client._request_stream(self, stream, *args, **kwargs)
     76 def _request_stream(
     77   self,
     78   *args,
     79   stream: bool = False,
     80   **kwargs,
     81 ) -> Union[Mapping[str, Any], Iterator[Mapping[str, Any]]]:
---> 82   return self._stream(*args, **kwargs) if stream else self._request(*args, **kwargs).json()

File /usr/local/lib/python3.11/site-packages/ollama/_client.py:58, in Client._request(self, method, url, **kwargs)
     56   response.raise_for_status()
     57 except httpx.HTTPStatusError as e:
---> 58   raise ResponseError(e.response.text, e.response.status_code) from None
     60 return response

ResponseError: "
}

**Comment by logan-markewich** [comment]
Was there no response error? Or did the copy-paste fail? Kind of hard to debug without knowing at least the response code haha

**Comment by hatianzhang** [comment]
@AutoAPEAI did you install llava in ollama?
`ollama run llava` run this first to try?

**Comment by AutoAPEAI** [comment]
> Was there no response error? Or did the copy-paste fail? Kind of hard to debug without knowing at least the response code haha

Hey, I have updated the log and the error name is ResponseError:

"name": "ResponseError",
"message": "",
"stack": "---------------------------------------------------------------------------
ResponseError Traceback (most recent call last)
Cell In[13], line 1
----> 1 response = mm_program(query_str="Can you summarize what is in the image?")
2 for res in response:
3 print(res)

File /usr/local/lib/python3.11/site-packages/llama_index/program/multi_modal_llm_program.py:85, in MultiModalLLMCompletionProgram.call(self, llm_kwargs, *args, **kwargs)

**Comment by AutoAPEAI** [comment]
> @AutoAPEAI did you install llava in ollama? `ollama run llava` run this first to try?
> 
> I cannot reproduce

Hey, llava was installed. And the terminal command runs perfectly:

ollama run llava
>>> hey
 Hello! How can I help you today? 

>>> Send a message (/? for help)

**Comment by logan-markewich** [comment]
@AutoAPEAI that is not the full traceback.

Is it a connection refused error? A timeout error?

**Comment by logan-markewich** [comment]
Assuming you started ollama with `ollama serve`, I'm not sure what the issue is. I cannot reproduce.... my guess is you need to increase the request_timeout

**Comment by AutoAPEAI** [comment]
> @AutoAPEAI that is not the full traceback.
> 
> Is it a connection refused error? A timeout error? I feel like there should be more info, or no?

the complete log is in my original post, i pasted here again:

{
	"name": "ResponseError",
	"message": "",
	"stack": "---------------------------------------------------------------------------
ResponseError                             Traceback (most recent call last)
Cell In[7], line 1
----> 1 response = mm_program(query_str=\"Can you summarize what is in the image?\")
      2 for res in response:
      3     print(res)

File /usr/local/lib/python3.11/site-packages/llama_index/program/multi_modal_llm_program.py:85, in MultiModalLLMCompletionProgram.__call__(self, llm_kwargs, *args, **kwargs)
     82 llm_kwargs = llm_kwargs or {}
     83 formatted_prompt = self._prompt.format(llm=self._multi_modal_llm, **kwargs)
---> 85 response = self._multi_modal_llm.complete(
     86     formatted_prompt,
     87     image_documents=self._image_documents,
     88     **llm_kwargs,
     89 )
     91 raw_output = response.text
     92 if self._verbose:

File /usr/local/lib/python3.11/site-packages/llama_index/multi_modal_llms/ollama.py:158, in OllamaMultiModal.complete(self, prompt, image_documents, formatted, **kwargs)
    155 \"\"\"Complete.\"\"\"
    156 import ollama
--> 158 response = ollama.generate(
    159     model=self.model,
    160     prompt=prompt,
    161     images=image_documents_to_base64(image_documents),
    162     stream=False,
    163     options=self._model_kwargs,
    164 )
    165 return CompletionResponse(
    166     text=response[\"response\"],
    167     raw=response,
    168     additional_kwargs=get_additional_kwargs(response, (\"response\",)),
    169 )

File /usr/local/lib/python3.11/site-packages/ollama/_client.py:111, in Client.generate(self, model, prompt, system, template, context, stream, raw, format, images, options, keep_alive)
    108 if not model:
    109   raise RequestError('must provide a model')
--> 111 return self._request_stream(
    112   'POST',
    113   '/api/generate',
    114   json={
    115     'model': model,
    116     'prompt': prompt,
    117     'system': system,
    118     'template': template,
    119     'context': context or [],
    120     'stream': stream,
    121     'raw': raw,
    122     'images': [_encode_image(image) for image in images or []],
    123     'format': format,
    124     'options': options or {},
    125     'keep_alive': keep_alive,
    126   },
    127   stream=stream,
    128 )

File /usr/local/lib/python3.11/site-packages/ollama/_client.py:82, in Client._request_stream(self, stream, *args, **kwargs)
     76 def _request_stream(
     77   self,
     78   *args,
     79   stream: bool = False,
     80   **kwargs,
     81 ) -> Union[Mapping[str, Any], Iterator[Mapping[str, Any]]]:
---> 82   return self._stream(*args, **kwargs) if stream else self._request(*args, **kwargs).json()

File /usr/local/lib/python3.11/site-packages/ollama/_client.py:58, in Client._request(self, method, url, **kwargs)
     56   response.raise_for_status()
     57 except httpx.HTTPStatusError as e:
---> 58   raise ResponseError(e.response.text, e.response.status_code) from None
     60 return response

ResponseError: "
}

**Comment by AutoAPEAI** [comment]
> > @AutoAPEAI that is not the full traceback.
> > Is it a connection refused error? A timeout error? I feel like there should be more info, or no?
> 
> the complete log is in my original post, i pasted here again:
> 
> { "name": "ResponseError", "message": "", "stack": "--------------------------------------------------------------------------- ResponseError Traceback (most recent call last) Cell In[7], line 1 ----> 1 response = mm_program(query_str="Can you summarize what is in the image?") 2 for res in response: 3 print(res)
> 
> File /usr/local/lib/python3.11/site-packages/llama_index/program/multi_modal_llm_program.py:85, in MultiModalLLMCompletionProgram.**call**(self, llm_kwargs, *args, **kwargs) 82 llm_kwargs = llm_kwargs or {} 83 formatted_prompt = self._prompt.format(llm=self._multi_modal_llm, **kwargs) ---> 85 response = self._multi_modal_llm.complete( 86 formatted_prompt, 87 image_documents=self._image_documents, 88 **llm_kwargs, 89 ) 91 raw_output = response.text 92 if self._verbose:
> 
> File /usr/local/lib/python3.11/site-packages/llama_index/multi_modal_llms/ollama.py:158, in OllamaMultiModal.complete(self, prompt, image_documents, formatted, **kwargs) 155 """Complete.""" 156 import ollama --> 158 response = ollama.generate( 159 model=self.model, 160 prompt=prompt, 161 images=image_documents_to_base64(image_documents), 162 stream=False, 163 options=self._model_kwargs, 164 ) 165 return CompletionResponse( 166 text=response["response"], 167 raw=response, 168 additional_kwargs=get_additional_kwargs(response, ("response",)), 169 )
> 
> File /usr/local/lib/python3.11/site-packages/ollama/_client.py:111, in Client.generate(self, model, prompt, system, template, context, stream, raw, format, images, options, keep_alive) 108 if not model: 109 raise RequestError('must provide a model') --> 111 return self._request_stream( 112 'POST', 113 '/api/generate', 114 json={ 115 'model': model, 116 'prompt': prompt, 117 'system': system, 118 'template': template, 119 'context': context or [], 120 'stream': stream, 121 'raw': raw, 122 'images': [_encode_image(image) for image in images or []], 123 'format': format, 124 'options': options or {}, 125 'keep_alive': keep_alive, 126 }, 127 stream=stream, 128 )
> 
> File /usr/local/lib/python3.11/site-packages/ollama/_client.py:82, in Client._request_stream(self, stream, *args, **kwargs) 76 def _request_stream( 77 self, 78 *args, 79 stream: bool = False, 80 **kwargs, 81 ) -> Union[Mapping[str, Any], Iterator[Mapping[str, Any]]]: ---> 82 return self._stream(*args, **kwargs) if stream else self._request(*args, **kwargs).json()
> 
> File /usr/local/lib/python3.11/site-packages/ollama/_client.py:58, in Client._request(self, method, url, **kwargs) 56 response.raise_for_status() 57 except httpx.HTTPStatusError as e: ---> 58 raise ResponseError(e.response.text, e.response.status_code) from None 60 return response
> 
> ResponseError: " }

and at the same time, the curl command works fine:

!curl http://localhost:11434/api/generate -d '{"model": "llava","prompt":"hello"}'

{"model":"llava","created_at":"2024-02-11T02:59:51.020361Z","response":" Hello","done":false}
{"model":"llava","created_at":"2024-02-11T02:59:51.252565Z","response":"!","done":false}
{"model":"llava","created_at":"2024-02-11T02:59:51.4862Z","response":" How","done":false}
{"model":"llava","created_at":"2024-02-11T02:59:51.726647Z","response":" can","done":false}
{"model":"llava","created_at":"2024-02-11T02:59:52.045212Z","response":" I","done":false}
{"model":"llava","created_at":"2024-02-11T02:59:52.301685Z","response":" help","done":false}
{"model":"llava","created_at":"2024-02-11T02:59:52.542726Z","response":" you","done":false}
{"model":"llava","created_at":"2024-02-11T02:59:52.816846Z","response":" today","done":false}
{"model":"llava","created_at":"2024-02-11T02:59:53.124538Z","response":"?","done":false}
{"model":"llava","created_at":"2024-02-11T02:59:53.370675Z","response":" ","done":false}
{"model":"llava","created_at":"2024-02-11T02:59:53.625923Z","response":"","done":true,"context":[733,16289,28793,6312,28709,733,28748,16289,28793,22557,28808,1602,541,315,1316,368,3154,28804,28705],"total_duration":15861320519,"load_duration":11981360163,"prompt_eval_count":10,"prompt_eval_duration":1272667000,"eval_count":11,"eval_duration":2605466000}

**Comment by logan-markewich** [comment]
RequestError('must provide a model')

How did you setup your ollama llm?

**Comment by abaranovskis-redsamurai** [comment]
It seems like Ollama multimodal LlamaIndex docs are removed? Any reason for this?

**Comment by logan-markewich** [comment]
@abaranovskis-redsamurai seems like a small issue when we merged the 2M line pr for v0.10 lol

We can add it back

**Comment by hatianzhang** [comment]
@logan-markewich I can add it back

**Comment by hatianzhang** [comment]
PR https://github.com/run-llama/llama_index/pull/11253

**Comment by SuperEllipse** [comment]
SOLVED: I had this issue too, I am using OLLAMA in a dockerized container. I fixed this by setting the base_url property of Ollama object, since in my company setup we could only use certain ports for local host. Hope this helps. Here is the code i used.. I am setting the OLLAMA_HOST value in my docker environment variable
    model = Ollama(model="gemma")
    model.base_url = f'http://{os.getenv("OLLAMA_HOST")}'
    print(model.base_url)
